File: ActionUIAndroid/library/src/main/java/com/abracode/actionui/Common/ActionUIElement.kt
package com.abracode.actionui.Common

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject

/**
 * Base shape of a decoded ActionUI element.
 *
 * ## Named containers
 *
 * SwiftUI's `ActionUIElement` stores every nested view under a stringly-typed
 * `subviews: [String: Any]?` dictionary whose keys are *named containers* - some
 * holding an **array** of elements (`children`, `destinations`, `toolbar`,
 * `commands`, and `rows` as an array-of-arrays) and some a **single** child
 * element (`content`, `destination`, `sidebar`, `detail`, `label`, `popover`,
 * `template`, `sheet`, `fullScreenCover`, `overlay`, `background`). See
 * `ActionUI/Common/ActionUIElement.swift`.
 *
 * Android models these as explicit, typed fields rather than one untyped map:
 * it is idiomatic Kotlin, decodes natively with `kotlinx.serialization` (a JSON
 * array -> `List<ActionUIElement>`, a JSON object -> `ActionUIElement`), and the
 * existing renderer already reads `element.children` directly everywhere. New
 * named containers are added here one field at a time, as the elements that use
 * them are ported - the same incremental stance the rest of the port takes.
 *
 * Implemented so far:
 *   * [children] - the array container (stacks, Group, ...).
 *   * [content]  - the first **single-child** container (ScrollView, and the
 *     GroupBox / LabeledContent / DisclosureGroup family). A single child,
 *     never an array on Android: author multiple children by wrapping them in a
 *     `VStack` / `Group`.
 *   * [template] - the data-driven repeater container (`List`, `Section`): a
 *     single element rendered once per data row, with `$1`/`$2`/`$0` column
 *     substitution (see `Helpers/TemplateHelper.kt`). Unlike [children] /
 *     [content] it is **excluded from [subElements]**: a template is instanced
 *     per row through throw-away rendering, not registered once in the window's
 *     [ViewModel] pool, so its own id (and its descendants') must not collide
 *     with the rest of the tree. Mirrors Swift, where template instances use
 *     throw-away `ViewModel`s rather than the window pool.
 *   * [destination] - a `NavigationLink`'s inline push target (a single view).
 *   * [destinations] - a `NavigationStack`'s array of push targets, each
 *     addressed by its `id` via a link/row's `destinationViewId`. Both
 *     navigation containers are registered (included in [subElements]) so their
 *     descendants get [ViewModel]s and are host-addressable.
 *   * [sidebar] / [detail] - a `NavigationSplitView`'s two single-child panes
 *     (the list pane and the detail pane). Registered (in [subElements]); the
 *     detail pane is driven by the sidebar selection (`destinationViewId` ->
 *     `destinations`) on tablets/foldables, collapsing to one pane on phones.
 *   * [toolbar] - the array of `ToolbarItem`/`ToolbarItemGroup` declared on any
 *     element; consumed by the navigation screen's `ToolbarHost` (Scaffold top/
 *     bottom bar), not rendered inline. Registered (in [subElements]) so each
 *     item's `content`/`children` get [ViewModel]s.
 *   * [sheet] / [fullScreenCover] - the element-level modal subviews any element
 *     may declare; presented by `Helpers/ElementModalHost.kt` when the carrier's
 *     `states["sheetVisible"]` / `states["fullScreenCoverVisible"]` flips true,
 *     not rendered inline. Registered (in [subElements]) so the modal content's
 *     controls get [ViewModel]s and are host-addressable, like on Apple.
 *   * [rows] - `Grid`'s array-of-arrays container: each inner array is one
 *     grid row's cells, in column order. Registered (flattened into
 *     [subElements]) so cells get [ViewModel]s like any other child.
 *
 * Any traversal of the element tree (id registration, etc.) should iterate
 * [subElements] so it automatically covers every named container as more are
 * added (the per-row [template] excepted, by design).
 */
interface ActionUIElementBase {
    val id: Int
    val type: String
    val properties: JsonObject?
    val children: List<ActionUIElement>?
    val content: ActionUIElement?
    val template: ActionUIElement?
    val destination: ActionUIElement?
    val destinations: List<ActionUIElement>?
    val sidebar: ActionUIElement?
    val detail: ActionUIElement?
    val toolbar: List<ActionUIElement>?
    val sheet: ActionUIElement?
    val fullScreenCover: ActionUIElement?
    val rows: List<List<ActionUIElement>>?
}

@Serializable
data class ActionUIElement(
    override val id: Int = 0,
    override val type: String = "",
    override val properties: JsonObject? = null,
    override val children: List<ActionUIElement>? = null,
    override val content: ActionUIElement? = null,
    override val template: ActionUIElement? = null,
    override val destination: ActionUIElement? = null,
    override val destinations: List<ActionUIElement>? = null,
    override val sidebar: ActionUIElement? = null,
    override val detail: ActionUIElement? = null,
    override val toolbar: List<ActionUIElement>? = null,
    override val sheet: ActionUIElement? = null,
    override val fullScreenCover: ActionUIElement? = null,
    override val rows: List<List<ActionUIElement>>? = null
) : ActionUIElementBase

/**
 * Every directly-contained child element, flattened across all named containers
 * in a stable order, for tree traversal (e.g. seeding view models by id). Add
 * each new single-child / array container here so existing walks pick it up
 * without change. The [template] container is intentionally **not** included:
 * its instances are rendered per row and never registered in the window pool.
 */
fun ActionUIElement.subElements(): List<ActionUIElement> = buildList {
    children?.let { addAll(it) }
    content?.let { add(it) }
    destination?.let { add(it) }
    destinations?.let { addAll(it) }
    sidebar?.let { add(it) }
    detail?.let { add(it) }
    toolbar?.let { addAll(it) }
    sheet?.let { add(it) }
    fullScreenCover?.let { add(it) }
    rows?.forEach { addAll(it) }
}

File: ActionUIWeb/src/Common/ActionUIElement.js
// ActionUIElement.js — parse raw JSON into the element tree.
// Web analog of ActionUI/Common/ActionUIElement.swift (PoC subset).
//
// {
//   "type": "VStack",
//   "id": 1,              // Optional: positive integer for programmatic interaction
//   "properties": {},     // Optional: view-specific properties
//   "children": []        // Optional: child elements (stored in subviews.children)
// }
//
// Subset routed so far: the "children" and "destinations" (arrays); "content",
// "label", "template", "sidebar", "detail" and "destination" (single); plus "rows"
// (array-of-arrays, for Grid's GridRows). "label" is the custom-trigger view
// (Menu's label-instead-of-title; later Button/Toggle). "template" is the
// data-driven repeater's per-row prototype (List's template mode; later Section).
// "sidebar"/"content"/"detail" are NavigationSplitView's panes and "destinations"
// its detail targets; "destination" is NavigationLink's inline push target;
// "toolbar" is the array of ToolbarItem/ToolbarItemGroup chrome carriers any view
// may declare. "sheet"/"fullScreenCover"/"popover" are the element-level
// presentation modifiers any view may carry (Helpers/PresentationModifier.js).
// "overlay"/"background" are the view-valued decoration subviews any view may
// carry (Helpers/DecorationModifier.js) — drawn over / behind the element without
// affecting its layout size. Note "background" is also a string *property* (a
// color); the subview form (an object) is the decoration view, and the two can
// coexist (the color paints in front of the background view, per Apple).

let negativeIDCounter = -1;

// "contextMenu" is an array subview (the menu's action items) and "contextMenuPreview"
// a single one (the optional preview view) — mirroring SwiftUI's two-builder
// `.contextMenu(menuItems:preview:)`. Modeled as element subtrees so a designer can
// place real elements (an Image preview, an HStack), not a button-descriptor array.
// "swipeActions" is likewise an array subview (the swipe action Buttons) — SwiftUI's
// `.swipeActions(edge:allowsFullSwipe:)`, a ViewBuilder of real Button elements.
// "persistentToolbar" is the same ToolbarItem/ToolbarItemGroup shape as "toolbar", but
// declared on a NavigationStack / NavigationSplitView: its items stay in the bar on every
// screen inside that container instead of on the one screen that declares them. Only those
// two types read it (Views/NavigationStack.js, Views/NavigationSplitView.js); anywhere else
// it renders nothing and Window.fromJSON warns.
const SUBVIEW_ARRAY_KEYS = ["children", "destinations", "toolbar", "persistentToolbar", "contextMenu", "swipeActions"];
const SUBVIEW_SINGLE_KEYS = ["content", "label", "template", "sidebar", "detail", "destination",
    "sheet", "fullScreenCover", "popover", "overlay", "background", "contextMenuPreview", "safeAreaInset"];
// Keys whose value is an array of arrays of elements (Grid's "rows": one inner
// array per GridRow). Stored as [[ActionUIElement]], mirroring Grid.swift's
// `subviews["rows"] as? [[any ActionUIElementBase]]`.
const SUBVIEW_NESTED_ARRAY_KEYS = ["rows"];

// The element types that own a navigation context rather than being a single screen. A
// toolbar authored on one of these belongs to the CONTAINER, not to the screen it happens
// to sit next to. Defined here, in the model, because both the parser (below) and the
// renderer (Helpers/ToolbarHelper.js) have to agree on the list and the model is the layer
// neither of them can avoid importing.
export const NAVIGATION_CONTAINER_TYPES = new Set(["NavigationStack", "NavigationSplitView"]);

export function isNavigationContainer(type) {
    return NAVIGATION_CONTAINER_TYPES.has(type);
}

export class ActionUIElement {
    constructor(id, type, properties, subviews) {
        this.id = id;
        this.type = type;
        this.properties = properties;
        this.subviews = subviews; // null or { children: [ActionUIElement], content: ActionUIElement }
    }

    static fromObject(raw, logger) {
        if (typeof raw !== "object" || raw === null || Array.isArray(raw)) {
            logger.log(`Element must be a JSON object, got ${JSON.stringify(raw)}`, "error");
            return null;
        }
        if (typeof raw.type !== "string" || raw.type.length === 0) {
            logger.log(`Element missing required "type" string`, "error");
            return null;
        }
        const id = Number.isInteger(raw.id) ? raw.id : negativeIDCounter--;
        const properties = (typeof raw.properties === "object" && raw.properties !== null)
            ? raw.properties : {};

        let subviews = null;
        for (const key of SUBVIEW_ARRAY_KEYS) {
            if (Array.isArray(raw[key])) {
                const children = raw[key]
                    .map((child) => ActionUIElement.fromObject(child, logger))
                    .filter((child) => child !== null);
                subviews = subviews ?? {};
                subviews[key] = children;
            }
        }
        for (const key of SUBVIEW_SINGLE_KEYS) {
            if (typeof raw[key] === "object" && raw[key] !== null) {
                const child = ActionUIElement.fromObject(raw[key], logger);
                if (child !== null) {
                    subviews = subviews ?? {};
                    subviews[key] = child;
                }
            }
        }
        for (const key of SUBVIEW_NESTED_ARRAY_KEYS) {
            if (Array.isArray(raw[key])) {
                const rows = raw[key]
                    .filter((row) => Array.isArray(row))
                    .map((row) => row
                        .map((child) => ActionUIElement.fromObject(child, logger))
                        .filter((child) => child !== null));
                subviews = subviews ?? {};
                subviews[key] = rows;
            }
        }
        // The two ways a toolbar key can be authored where nothing will read it. Reported
        // here because fromObject visits every element exactly once, on EVERY parse path -
        // the window document, a LoadableView, a modal description, a runtime insertion -
        // so a misauthored sub-document says so too. The renderer stays silent: by build
        // time the key has already been folded into the container's items or dropped, and
        // a render-time warning would repeat on every pane switch.
        const container = NAVIGATION_CONTAINER_TYPES.has(raw.type);
        if (container && subviews?.toolbar?.length) {
            logger.log(
                `Deprecated: 'toolbar' on ${raw.type} (id ${id}) is treated as 'persistentToolbar' - its items `
                + "stay in the bar on every screen inside the container. Rename the key to silence this; declare "
                + "a screen's own toolbar on that screen's element instead.",
                "warning",
            );
        }
        if (!container && subviews?.persistentToolbar?.length) {
            logger.log(
                `'persistentToolbar' on ${raw.type} (id ${id}) is ignored - only NavigationStack and `
                + "NavigationSplitView have screens to keep it on. Use 'toolbar' for this view's own toolbar, or "
                + "move the key to the enclosing navigation container.",
                "warning",
            );
        }

        return new ActionUIElement(id, raw.type, properties, subviews);
    }

    static fromJSONString(jsonString, logger) {
        let raw;
        try {
            raw = JSON.parse(jsonString);
        } catch (error) {
            logger.log(`Invalid JSON: ${error.message}`, "error");
            return null;
        }
        return ActionUIElement.fromObject(raw, logger);
    }

    children() {
        return this.subviews?.children ?? [];
    }

    // The data-driven repeater's per-row prototype (List's template mode), or
    // null. Never built directly — the owner substitutes a copy per data row.
    template() {
        return this.subviews?.template ?? null;
    }

    // NavigationSplitView's panes and detail targets. sidebar() / content() /
    // detail() are single child elements (or null); destinations() is the array
    // of detail targets addressed by id (empty when absent).
    sidebar() {
        return this.subviews?.sidebar ?? null;
    }

    content() {
        return this.subviews?.content ?? null;
    }

    detail() {
        return this.subviews?.detail ?? null;
    }

    destinations() {
        return this.subviews?.destinations ?? [];
    }

    // NavigationLink's inline push target (Form 1), or null. Hoisted by id into
    // the enclosing NavigationStack's destination registry (it must carry an id
    // to be addressable, like Android's NavHost routes).
    destination() {
        return this.subviews?.destination ?? null;
    }

    // The view's toolbar chrome carriers (ToolbarItem / ToolbarItemGroup), or an
    // empty array. Rendered as screen chrome by ToolbarHelper, not built directly.
    toolbar() {
        return this.subviews?.toolbar ?? [];
    }

    // A navigation container's persistent chrome carriers: the items it keeps in the bar
    // on every screen inside it. Read through Helpers/ToolbarHelper.js's
    // persistentToolbarItems(), which also folds in the deprecated `toolbar` spelling.
    persistentToolbar() {
        return this.subviews?.persistentToolbar ?? [];
    }
}
